### ISSUE TITLE
False Positive UnnecessarySafeCalls re: Android Views (1.15.0)

### ISSUE THREAD

**Issue by drinkthestars** [detekt/detekt#3409]
## Expected Behavior
Elvis operators `?` for safe access of Android `View`s must not be falsely reported as unnecessary.

## Observed Behavior
False positive `UnnecessarySafeCall`s were reported when using `?` operator on nullable Android `View`s. For example:

``` Kotlin
viewBinding.viewStubProxy.viewStub?.layoutResource = R.layout.some_layout
```
Where `viewStub` refers to the following Java getter (return value marked as `@Nullable`):

``` Java
    @Nullable
    public ViewStub getViewStub() {
        return mViewStub;
    }
```

## Your Environment
* Version of detekt used: `1.15.0`
* Version of Gradle used (if applicable): `6.7.1`
* Version of Kotlin used: `1.3.72`
* Operating System and version: `macOS Catalina 10.15.7`

**Comment by schalkms** [comment]
We use official Kotlin compiler diagnostics in this rule for detecting the unnecessary safe call. Please see here for more information: https://github.com/detekt/detekt/blob/27303367cc351290ec43af1af672a15cd58f1810/detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/UnnecessarySafeCall.kt#L42-L61

If you think this is an issue, I guess this should be forwarded to the compiler team.

**Comment by BraisGabin** [comment]
I need to do more research but today one collegue at work reported me a false positive in UnnecesarySafeCall related with android views too. And AndroidStudio was not complaining about it.

As I said I didn't have the time to check it but I think that this issue is related with ViewBinding. ViewBinding is a Kotlin compiler plugin so we don't have the information of that class in the analysis (as detekt make it clear with a lot of noise in the console).

I assume that @drinkthestars is using ViewBinding too so that can be the problem. We don't know the type of `viewStubProxy` so we can't know that `viewStub` can be `null`.

**Comment by BraisGabin** [comment]
I don't know how should we proceed here:
- The false positives are really annoying. And they make the user lose trust in detekt. So we should address them.
- The Kotlin compiler plugins are getting popular so we should support them. We get that support by using https://github.com/detekt/detekt-compiler-plugin (correct me if I'm wrong here) but it's not production ready yet (right?)
- I don't know how much complexity this brings us to the rules. Is it worth it to fix this kind of issues? Should we wait for detekt-compiler-plugin?

In my opinion we should fix this kind of issues because the first point is really important.

**Comment by drinkthestars** [comment]
@BraisGabin correct, I have seen this when using `ViewBinding` (and data binding). I suspect your hunch is correct and that this happens with `ViewBinding` and related components, in that case https://github.com/detekt/detekt-compiler-plugin might be the best bet.

To circumvent at the moment, I'm `@Suppress`ing them. _However_, I don't recall running into this particular false positive prior to detekt version `1.15.0` though.

**Comment by mhernand40** [comment]
I'm actually seeing yet another `UnnecessarySafeCall` false positive that involves Kotlin extension functions that talk to Java-generated Protobufs from the [Protobuf Gradle Plugin](https://github.com/google/protobuf-gradle-plugin) within an Android library module.

Sorry, I do not have a small project with the isolated repro at the moment but here are some snippets of what a reduced repro looks like:

**Protobuf Source:**
```protobuf
message User {
    optional Name name = 1;

    message Name {
        optional string first = 1;
        optional string last = 2;
    }
}
```

**Kotlin Source:**
```Kotlin
data class User(val name: Name) {
    data class Name(val first: String, val last: String)
}

// DetektMain complains twice with UnnecessarySafeCall - [toUserOrNullSingleLine]
fun TinderProto.User.toUserOrNullSingleLine(): User? {
    return nameOrNull()?.toName()?.let { User(name = it) }
}

// DetektMain does NOT complain when safe calls are broken into multiple lines
fun TinderProto.User.toUserOrNullMultilineLine(): User? {
    return nameOrNull()
        ?.toName()
        ?.let { User(name = it) }
}

private fun Proto.User.nameOrNull(): Proto.User.Name? {
    return if (hasName()) name else null
}

private fun Proto.User.Name.toName(): User.Name {
    return User.Name(first = first, last = last)
}
```

Notice how in this case, one can "work around" this issue by breaking up the safe call chaining into separate lines. I suspect the issue is related to a combination of Kotlin talking to Java code gen all within an Android module.

**Comment by schalkms** [comment]
> I think that this issue is related with ViewBinding. ViewBinding is a Kotlin compiler plugin so we don't have the information of that class in the analysis (as detekt make it clear with a lot of noise in the console).

Thanks for the valuable feedback guys! This issue is solved, if the compiler plugin solution is pushed forward. Without having the detekt compiler plugin in use, I guess, this issue won’t be resolved.

**Comment by cortinico** [comment]
> I think we need to understand where is the problem with a real scenario.

After some investigation on #3414, I believe the culprit is in the classpath used for type resolution. See #3419 description for a better context on it.

I'd be curious to understand which task are you invoking @drinkthestars once you're noticing this. Is it a `detektMain`, `detektDebug`, or you have your own custom task?

**Comment by BraisGabin** [comment]
> I'm unsure on how detekt-compiler-plugin would solve the issue here. What we're doing is still delegating to a Kotlin compiler inspection.

But we will have the real compiler classpath so, if it compiles, we will have the view binding classes, right?

**Comment by mhernand40** [comment]
@cortinico is a reproducer project still required now that https://github.com/detekt/detekt/pull/3419 is up for review?

**Comment by drinkthestars** [comment]
@cortinico 

>I'd be curious to understand which task are you invoking @drinkthestars once you're noticing this. Is it a detektMain, detektDebug, or you have your own custom task?

It was `detektMain`.

**Comment by cortinico** [comment]
> is a reproducer project still required

Probably not as we were able to isolate a test case to reproduce the issue.

> It was detektMain.

That's interesting. I think the issue in this case could be related to the ViewBinding feature.
